### PR TITLE
Replace `<a>` and `<button>` with `<PlainButton>`

### PR DESCRIPTION
**Pull request by rafawendel**
## What type of PR is this? (check all applicable)

- [X] Feature

## Description
This PR is a split on #5409
This PR follows on #5418, #5419, #5420 and #5424

It is the fourth step to fix issues on standardization, interactivity and improve keyboard accessibility.

This aims to clean most remaining `<a>` and `<button>` and replace it with own semantic buttons. There should be no visual changes.

### The following behavior for links, buttons, link-like and button-like components is aimed:

- Links that look like links:  `<Link>` - Renders `<a>`
- Buttons that look like buttons: `<Button>` - Renders Antd `<Button>`
- Links that look like buttons:  `<Link.Button>` - Renders Antd `<Button>`, which is rendered as an `<a>`
- Text Buttons "buttons that look like links": `<PlainButton>` - Renders styleless `<button>`

## Major visual changes

**Before**

https://user-images.githubusercontent.com/44540213/113716056-90abd680-96c0-11eb-925d-3cf2ccfd2159.mp4

**After**

https://user-images.githubusercontent.com/44540213/113715916-69550980-96c0-11eb-91c7-886d34231114.mp4

### CODE REVIEW

File: client/app/components/queries/SchemaBrowser.jsx [hunk 1]
```diff
@@ -46,20 +46,18 @@ function SchemaItem({ item, expanded, onToggle, onSelect, ...props }) {
 
   return (
     <div {...props}>
-      {/* TODO: Replace with a button */}
-      <div className="table-name" onClick={onToggle}>
+      <PlainButton className="table-name" onClick={onToggle}>
```
[the remark this answers is not in the document]
Comment (rafawendel) [reply client/app/components/queries/SchemaBrowser.jsx:49]: Lol

File: client/app/components/empty-state/EmptyState.jsx [hunk 2]
```diff
@@ -17,9 +19,9 @@ export function Step({ show, completed, text, url, urlTarget, urlText, onClick }
 
   return (
     <li className={classNames({ done: completed })}>
-      <Link href={url} onClick={onClick} target={urlTarget}>
+      <Button type="link" href={url} onClick={onClick} target={urlTarget}>
```
[the remark this answers is not in the document]
Comment (rafawendel) [reply client/app/components/empty-state/EmptyState.jsx:22]: `<Link>` _should_ throw a type error if it get's a click handler. This was sort of a workaround. I'm trying to think in a more general solution to this.

File: client/app/components/ApplicationArea/ApplicationLayout/DesktopNavbar.jsx [hunk 3]
```diff
@@ -129,9 +130,9 @@ export default function DesktopNavbar() {
             )}
             {canCreateDashboard && (
               <Menu.Item key="new-dashboard">
-                <a data-test="CreateDashboardMenuItem" onMouseUp={() => CreateDashboardDialog.showModal()}>
+                <PlainButton data-test="CreateDashboardMenuItem" onClick={() => CreateDashboardDialog.showModal()}>
```
[the remark this answers is not in the document]
Comment (rafawendel) [reply client/app/components/ApplicationArea/ApplicationLayout/DesktopNavbar.jsx:133]: I've tested it, works fine. Both with mouse and focus. Do you have an idea of another problem that could be related?

[the remark this answers is not in the document]
Comment (kravets-levko) [reply client/app/components/ApplicationArea/ApplicationLayout/DesktopNavbar.jsx:133]: Maybe that was something with older Antd version. If it works - I have no other questions 👍

File: client/app/components/empty-state/EmptyState.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (kravets-levko) [reply client/app/components/empty-state/EmptyState.jsx:22]: `<Link>` has some logic that allows to modify URL. Using `<Button>` here will break those links in Databricks internal fork

File: client/app/assets/less/inc/schema-browser.less [hunk 4]
```diff
@@ -1,102 +1,107 @@
-div.table-name {
```
Comment (thielium) [inline client/app/assets/less/inc/schema-browser.less:1]: This change feels bigger than the others. Can you summarize what the expected changes are to the schema browser?

File: client/app/components/InputWithCopy.jsx [hunk 5]
```diff
@@ -42,7 +43,10 @@ export default class InputWithCopy extends React.Component {
   render() {
     const copyButton = (
       <Tooltip title={this.state.copied || "Copy"}>
-        <CopyOutlinedIcon style={{ cursor: "pointer" }} onClick={this.copy} />
+        <PlainButton onClick={this.copy}>
+          {/* TODO: MISSING_VISUAL_FEEDBACK */}
```
Comment (thielium) [inline client/app/components/InputWithCopy.jsx:47]: Can we link this to an issue or provide more explanation?

File: client/app/components/Tooltip.tsx [hunk 6]
```diff
@@ -3,7 +3,7 @@ import AntTooltip, { TooltipProps } from "antd/lib/tooltip";
 import { isNil } from "lodash";
 
 export default function Tooltip({ title, ...restProps }: TooltipProps) {
-  const liveTitle = isNil(title) ? (
+  const liveTitle = !isNil(title) ? (
```
Comment (thielium) [inline client/app/components/Tooltip.tsx:6]: This feels far-reaching. Is this just a tweak to a recent change in logic? Is there any way we can add a test / screenshot or something to make sure this doesn't slip in in the future?

File: client/app/components/dashboards/dashboard-widget/VisualizationWidget.jsx [hunk 7]
```diff
@@ -180,18 +182,18 @@ function VisualizationWidgetFooter({ widget, isPublic, onRefresh, onExpand }) {
...
             onClick={() => refreshWidget(2)}>
             <i className={cx("zmdi zmdi-refresh", { "zmdi-hc-spin": refreshClickButtonId === 2 })} aria-hidden="true" />
             <span className="sr-only">
               {refreshClickButtonId === 2 ? "Refreshing, please wait." : "Press to refresh."}
             </span>
-          </a>
+          </PlainButton>
         )}
-        <a className="btn btn-sm btn-default hidden-print btn-transparent btn__refresh" onClick={onExpand}>
+        <PlainButton className="btn btn-sm btn-default hidden-print btn-transparent btn__refresh" onClick={onExpand}>
```
Comment (thielium) [inline client/app/components/dashboards/dashboard-widget/VisualizationWidget.jsx:194]: Any chance we can just use a normal button, here?

File: client/app/components/empty-state/EmptyState.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (thielium) [reply client/app/components/empty-state/EmptyState.jsx:22]: I do not agree that link should throw a type  error if it gets a click handler. 

It's valid to have a link perform one behavior when you click on it, and another if you middle click / open in a new tab (for example, if you want to open a modal, you don't want to refresh the page when you left click, but you do want a new page with the modal if you open a link to it in a new tab).

File: client/app/components/queries/AddToDashboardDialog.jsx [hunk 8]
```diff
@@ -89,7 +90,9 @@ function AddToDashboardDialog({ dialog, visualization }) {
           value={searchTerm}
           onChange={event => setSearchTerm(event.target.value)}
           suffix={
-            <CloseOutlinedIcon className={searchTerm === "" ? "hidden" : null} onClick={() => setSearchTerm("")} />
+            <PlainButton className={searchTerm === "" ? "hidden" : null} onClick={() => setSearchTerm("")}>
+              <CloseOutlinedIcon />
```
Comment (thielium) [inline client/app/components/queries/AddToDashboardDialog.jsx:94]: Do Icons not support click events / aria roles? Is this a styling change?

File: client/app/pages/alert/components/Query.jsx [hunk 9]
```diff
@@ -27,9 +28,9 @@ export default function QueryFormItem({ query, queryResult, onChange, editMode }
       <small>
         <WarningFilledIcon className="warning-icon-danger" /> This query has no <i>refresh schedule</i>.{" "}
         <Tooltip title="A query schedule is not necessary but is highly recommended for alerts. An Alert without a query schedule will only send notifications if a user in your organization manually executes this query.">
-          <a>
+          <PlainButton type="link">
```
Comment (thielium) [inline client/app/pages/alert/components/Query.jsx:31]: Is it correct to use a link `PlainButton` here? Don't they have a `role` of `button` associated when them?

File: client/app/pages/queries/QueryView.jsx [hunk 10]
```diff
@@ -122,10 +123,10 @@ function QueryView(props) {
             queryFlags.canEdit &&
             !addingDescription &&
             !fullscreen && (
-              <a className="label label-tag hidden-xs" role="none" onClick={() => setAddingDescription(true)}>
+              <PlainButton className="label label-tag hidden-xs" role="none" onClick={() => setAddingDescription(true)}>
```
Comment (thielium) [inline client/app/pages/queries/QueryView.jsx:126]: This also feels heavily styled. I'm not sure that heavily styled items should be wrapped in a `PlainButton` -- it feels like maybe just adding a `role` is appropriate?

File: client/app/assets/less/inc/schema-browser.less (hunk 4 above)
Comment (rafawendel) [reply client/app/assets/less/inc/schema-browser.less:1]: Changes expected are basically none. Just refactored the styles to make it easier to work with, as many of the positionings were done in a workaround-like manner. There were bugs as well: I've changed the way tooltips are positioned upon these components to make it a11y (and UI overall) friendly.

File: client/app/components/InputWithCopy.jsx (hunk 5 above)
Comment (rafawendel) [reply client/app/components/InputWithCopy.jsx:47]: We have the issue. The visual feedback here refers to focus/hover/active styles, which are lacking or inconsistent across the UI. [Tracker](https://databricks.atlassian.net/browse/REDASH-1102?atlOrigin=eyJpIjoiYTI1NWNjMTgyY2I2NGViMTk5MzMwMmRkZTk2MDA2ZGUiLCJwIjoiaiJ9)

File: client/app/components/Tooltip.tsx (hunk 6 above)
Comment (rafawendel) [reply client/app/components/Tooltip.tsx:6]: It is a bug. When I added the `isNil` I accidentally made all tooltips stop working.

File: client/app/components/dashboards/dashboard-widget/VisualizationWidget.jsx (hunk 7 above)
Comment (rafawendel) [reply client/app/components/dashboards/dashboard-widget/VisualizationWidget.jsx:194]: You mean a `<button>`? If that's the case, I'm trying to avoid it to keep consistency in the codebase. `<PlainButton>`'s only difference to it is the pointer.

File: client/app/components/queries/AddToDashboardDialog.jsx (hunk 8 above)
Comment (rafawendel) [reply client/app/components/queries/AddToDashboardDialog.jsx:94]: Wrapping in a button is better than native behavior for two reasons:

1. Out of the box keyboard support for `Enter` and `space`, according to the specs, without need for additional logic.
2. Better semantics: despite being a close icon, it is, after all, a close _button_, and this makes it explicit both in the code and for screen readers.
3. Extra one: only elements with interactive roles (which I surely could have added) can have `aria-label`s

File: client/app/pages/queries/QueryView.jsx (hunk 10 above)
Comment (rafawendel) [reply client/app/pages/queries/QueryView.jsx:126]: We end in the same discussion as [here](https://github.com/getredash/redash/pull/5433#discussion_r607374939)

File: client/app/pages/alert/components/Query.jsx (hunk 9 above)
Comment (rafawendel) [reply client/app/pages/alert/components/Query.jsx:31]: This `type` in `<PlainButton>` is only useful for styling the "text button" as an "Ant design link-looking text button". In this context it felt appropriate, as this is the standard followed in other parts of the codebase, e.g. `EmptyState`

File: client/app/components/InputWithCopy.jsx (hunk 5 above)
Comment (gabrieldutra) [reply client/app/components/InputWithCopy.jsx:47]: I keep my position against those TODO comments -> https://github.com/getredash/redash/pull/5424#discussion_r598001895

File: client/app/assets/less/inc/schema-browser.less (hunk 4 above)
Comment (rafawendel) [reply client/app/assets/less/inc/schema-browser.less:1]: Quick summary of the changes:

- Improved nesting
- Decoupled the button to open the drawer with the one that adds it to the editor
- Updated the box sizing to flexbox to achieve the same result more consistently
- Used flexbox to center, in place of paddings and margins
- Reduced repetition re-structuring how classes are set
- Added focus feedback
- Added transition
- TODO (not tracked): render of drawer has no transition

File: client/app/components/empty-state/EmptyState.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (rafawendel) [reply client/app/components/empty-state/EmptyState.jsx:22]: The problem in this case is for pieces of code like this: 
image

In practice, this component is sometimes a link and sometimes a button, which hurts its accessibility.

[the remark this answers is not in the document]
Comment (kravets-levko) [reply client/app/components/empty-state/EmptyState.jsx]: When `Step` component has a link - it can have either `urlText` + `onClick` or `urlText` + `url` + `urlTarget`. You can use this to decide what to render - link or button. Also, you could even update TS types to indicate this:

```ts
type StepCommonProps = { ... };

type StepLinkProps = {
  urlText: string;
  url: string;
  urlTarget: string;
}

type StepButtonProps = {
  urlText: string;
  onClick: ...;
}

type StepProps = StepCommonProps & (StepLinkProps | StepButtonProps);
```

In this case when you'll try to render `<Step>` component with `onClick` - TS will not allow you to use `url` and `urlTarget`, and if you'll add `url` and/or `urlTarget` - it will forbid `onClick`

[the remark this answers is not in the document]
Comment (rafawendel) [reply client/app/components/empty-state/EmptyState.jsx]: Update: see #5435

File: client/app/components/dashboards/dashboard-widget/VisualizationWidget.jsx (hunk 7 above)
Comment (thielium) [reply client/app/components/dashboards/dashboard-widget/VisualizationWidget.jsx:194]: I actually just meant a `Button`

File: client/app/components/queries/AddToDashboardDialog.jsx (hunk 8 above)
Comment (thielium) [reply client/app/components/queries/AddToDashboardDialog.jsx:94]: I think I buy it! 👍

File: client/app/components/empty-state/EmptyState.jsx [hunk 11]
```diff
@@ -12,16 +12,16 @@ import organizationStatus from "@/services/organizationStatus";
...
 
-export function Step({ show, completed, text, url, urlTarget, urlText, onClick }) {
+function Step({ show, completed, text, url, urlTarget, urlText, onClick }) {
   if (!show) {
     return null;
   }
 
   return (
     <li className={classNames({ done: completed })}>
-      <Button type="link" href={url} onClick={onClick} target={urlTarget}>
```
Comment (thielium) [inline client/app/components/empty-state/EmptyState.jsx]: Why `Link type="link"`?

**Review by thielium — APPROVED** [review]
Nothing blocking. If Percy's happy, so am I! Thanks for doing all this.
